### DISCUSSION TITLE
LoRA scheduled sampling setup for more stable no-prompt character voices

### DISCUSSION THREAD

**Discussion by TachibanaKimika** [fishaudio/fish-speech#1314] [Show and tell]
Fish Speech does include an official LoRA fine-tuning workflow. The [official docs](https://github.com/fishaudio/fish-speech/blob/main/docs/en/finetune.md) are also quite clear about its default behavior: it mainly learns speaking patterns rather than reliably learning the speaker's timbre, so inference still needs a Prompt. That matched my experience. After training with the default teacher-forcing setup, the character voice was hard to reproduce consistently without a Prompt.

I later adapted the scheduled sampling work from [thhyj/fish-audio-lora-full](https://gitee.com/thhyj/fish-audio-lora-full), brought the relevant changes onto Fish Speech commit `e5e2926`, and removed dataset-specific paths and private artifacts:

- [Fork](https://github.com/TachibanaKimika/fish-speech)
- [Implementation commit `166940c`](https://github.com/TachibanaKimika/fish-speech/commit/166940c301b8b4650b3110397a38bc68727a3afa)
- [Method notes](https://github.com/TachibanaKimika/fish-speech/blob/main/docs/scheduled-sampling-finetune.md)

Each training step now runs one or more no-gradient probes, samples the predicted semantic and acoustic codes, and mixes some of them back into the input. It then runs one gradient-enabled forward pass on the mixed context while keeping the original labels as the training target. The probe depth increases from 1 to 3 over the course of training, giving the model time to adjust to contexts that contain accumulated prediction errors.

There are also a few fixes that affect training stability. Codebook embedding scaling is now consistent between training and inference. Gradient checkpointing is non-reentrant so the slow-AR LoRA actually receives gradients, and document IDs keep packed utterances isolated from one another. These fixes affect the result alongside scheduled sampling, so this is not a clean ablation of scheduled sampling by itself.

This is the starting recipe I used:

```text
S2-Pro, LoRA r=8 / alpha=16
batch size 2, max length 1024, BF16
learning rate 5e-5, cosine schedule
SS_EPS=0.20, SS_WARMUP=2000
SS_DEPTH_MAX=3, SS_DEPTH_WARMUP=6000
SS_TEMP=0.70, SS_TOP_P=0.70, SS_TOP_K=30
TORCH_COMPILE=0
```

This setup ran for 10,000 steps on one RTX 4090 24 GB with a 4.6-hour Japanese single-speaker corpus. I saved a checkpoint every 500 steps and picked the model through fixed autoregressive listening tests. I did not simply use the final checkpoint or the one with the lowest validation loss.

The public samples compare three setups: base S2-Pro zero-shot with a fixed Prompt, LoRA with the same Prompt, and LoRA with no reference audio, reference text, or reference semantic tokens. In the [Kurumi example](https://huggingface.co/TachibanaKimika/fish-speech-lora/blob/main/adapters/tenshi_souzou_reboot/kurumi/README.md), their ERes2NetV2 SIM scores are `0.7564`, `0.8101`, and `0.7924` respectively.

Across six published adapters, prompted LoRA had higher SIM than base zero-shot in all six cases. No-prompt LoRA was higher in four of six, with a mean SIM of `0.7000` versus `0.6807` for prompted zero-shot. SIM only measures speaker timbre. It does not measure naturalness, emotion, acting, or text accuracy.

In practice, the selected LoRAs could reproduce the character timbre fairly consistently without a Prompt. They were more stable than my earlier teacher-forcing runs and often sounded better than base zero-shot. When a good reference recording was available, LoRA with a Prompt was usually still the best setup.

The full adapter collection and matched samples are available at [TachibanaKimika/fish-speech-lora](https://huggingface.co/TachibanaKimika/fish-speech-lora).

These results come from Japanese single-speaker datasets, so I would not treat this as a universal recipe for every language or dataset size. The 10,000 steps were an exploration budget for this experiment, not a recommended endpoint for every run.

---

*AI assistance disclosure: This post was drafted and edited with assistance from OpenAI Codex based on the linked code, experiment records, and samples.*
